hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -74,32 +100,362 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
+	}
+	return nil
+}
+
+func getNodesToUpgrade(nodes []*corev1.Node, targetConfig string, maxUnavailable int) []*corev1.Node {
+	// get unavailable machines
+	// In the MCO logic, unavailable is defined as any of:
+	// - config does not match
+	// - MCD is failing
+	// - Node is unscheduleable
+	// - NodeReady condition status is ConditionTrue,
+	// - NodeDiskPressure condition status is ConditionFalse,
+	// - NodeNetworkUnavailable condition status is ConditionFalse.
+	// TODO (jerzhang): consider what we want to do with node status here
+	// For now, we will just check current/desired config to see if any nodes is already updating
+	var numUnavailable int
+	for _, node := range nodes {
+		if node.Annotations[CurrentMachineConfigAnnotationKey] != node.Annotations[DesiredMachineConfigAnnotationKey] {
+			numUnavailable++
+		}
+	}
+
+	capacity := maxUnavailable - numUnavailable
+	// If we're at capacity, there's nothing to do.
+	if capacity < 1 {
 		return nil
 	}
+	// We only look at nodes which aren't already targeting our desired config
+	var candidateNodes []*corev1.Node
+	for _, node := range nodes {
+		if node.Annotations[DesiredMachineConfigAnnotationKey] != targetConfig {
+			candidateNodes = append(candidateNodes, node)
+		}
+	}
+
+	if len(candidateNodes) == 0 {
+		return nil
+	}
+
+	// Not sure if we need to order this
+	return candidateNodes[:capacity]
+}
+
+func (r *NodePoolReconciler) handleNodeDrainRequest(ctx context.Context, hostedClusterClient client.Client, node *corev1.Node, desiredState string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// TODO (jerzhang): delete the pod after we uncordon
+	// desiredVerb := strings.Split(desiredState, "-")[0]
+	// if desiredVerb == DrainerStateUncordon {
+	// }
+
+	// TODO (jerzhang): actually implement the node draining. For now, just set the singal and pretend we drained.
+	annotations := map[string]string{
+		LastAppliedDrainerAnnotationKey: desiredState,
+	}
+	if result, err := r.CreateOrUpdate(ctx, hostedClusterClient, node, func() error {
+		return r.reconcileNodeAnnotations(
+			ctx,
+			node,
+			annotations,
+		)
+	}); err != nil {
+		return fmt.Errorf("failed to reconcile node drain annotations: %w", err)
+	} else {
+		log.Info("Reconciled Node drain annotations", "result", result)
+	}
+	return nil
+}
+
+func (r *NodePoolReconciler) reconcileNodeAnnotations(ctx context.Context, node *corev1.Node, annotations map[string]string) error {
+	for k, v := range annotations {
+		node.Annotations[k] = v
+	}
+	return nil
+}
+
+func (r *NodePoolReconciler) reconcileInPlaceUpgradeManifests(ctx context.Context, hostedClusterClient client.Client, targetConfigVersionHash, ignEndpoint string, caCertBytes, tokenBytes []byte, nodePool *hyperv1.NodePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	namespace := inPlaceUpgradeNamespace(nodePool)
+
+	// Ensure the namespace exists and create if it isn't
+	if err := hostedClusterClient.Get(ctx, client.ObjectKeyFromObject(namespace), namespace); err != nil {
```

Comment:
could just use `r.CreateOrUpdate` here as well.

Reply:
I opted to just do this myself since there is nothing to mutate, and passing in a `nil` mutateFn would just cause the CreateOrUpdate in upsert to panic (as opposed to the controllerutil createorpatch which will instead do nothing without panicking). Is there a proper way to use CreateOrUpdate if I don't want to mutate anything?

Reply:
Will change to CreateOrUpdate for consistency then. I guess one last question pre-merge is, since the MCD PR has merged to 4.11, should we attempt to directly use the MCD image from the management cluster payload?

This would mean that 4.11 pre-MCD merge/4.10- would not work (and we would need to use updated nightlies for testing), but I guess that is the expectation anyway?